Revamped parallelism of parts of batch importer

tinwelint opened the pull request:
The main story here is composed of a couple of things:
- a new ForkedProcessorStep which does parallelization inside each
  batch, executed one by one. This to avoid difficulties parallelizing
  some steps which has a costly section which isn't parallelizable.
  With this new step items in a batch can be striped such that each
  forked processor knows which parts to process.
- better mechanical sympathy where most stages are optimized to work
  with batch sizes matching pages in the page cache of the store they
  (mainly) work with.
- more static node->relationship caches, initialised after knowing how
  many nodes have been imported. The static caches are faster than
  the dynamically growing.

The forked processor simplifies a couple of stages, there are now
no artificial additional steps for splitting or otherwise modify
batches to be better parallelizable. Also the whole stage scales
better with added processors because the old way of parallelizing
those stages often involved a step which was single-threaded and
acted as a divider-of-work. Such a step would often become the
bottleneck in the end anyway.

About mechanical sympathy the main problem previously was that reader
and writer of stages which read from and wrote to the same store actually
contended on each other. Given the smaller batch size, there were
multiple batches of read records for any given page. Later in the stage
where store was updated would often update the same page and so the
reader (still reading that page) would need to do mych more retry-
reading and so slow the whole stage down. Now with the aligned
batch sizes the reader doesn't contend with the writers in the
page cache.

Additionally the main store updating step have been split into steps
updating entities and properties separately, this to have the
entity updating able to go even faster.

The net result of this change as a whole should be that more often
the disk is the only main bottleneck. On test machines and development
laptops a 2x-3x performance improvement of the importer have
been observed.

changelog: [kernel] Faster import tool due to ability to better utilize all available resources and higher I/O saturation.

tinwelint committed:
Static cache arrays during import for better performance

Caches are instantiated later where maximum size is known
and so static arrays can be instantiated, they are faster than
the dynamically growing arrays and so the importer will be
faster as a whole.

The creation of these arrays still fallback on splitting between
on/off heap if it won't fit in either completely, so that
case will still perform that same. Best case will be better.

tinwelint committed:
More information when missing mandatory CSV header

tinwelint committed:
Parallel type checking in calculate dense node stage

so that it doesn't become a bottleneck

tinwelint committed:
Shutdown race issue in TicketedProcessing

tinwelint committed:
Doesn't remove processors if there are permits available

tinwelint committed:
Revamped parallelism of parts of batch importer

The main story here is composed of two things:

- a new ForkedProcessorStep which does parallelization inside each
  batch, executed one by one. This to avoid difficulties parallelizing
  some steps which has a costly section which isn't parallelizable.
  With this new step items in a batch can be striped such that each
  forked processor knows which parts to process.
- better mechanical sympathy where most stages are optimized to work
  with batch sizes matching pages in the page cache of the store they
  (mainly) work with.

The forked processor simplifies a couple of stages, there are now
no artificial additional steps for splitting or otherwise modify
batches to be better parallelizable. Also the whole stage scales
better with added processors because the old way of parallelizing
those stages often involved a step which was single-threaded and
acted as a divider-of-work. Such a step would often become the
bottleneck in the end anyway.

About mechanical sympathy the main problem previously was that reader
and writer of stages which read from and wrote to the same store actually
contended on each other. Given the smaller batch size, there were
multiple batches of read records for any given page. Later in the stage
where store was updated would often update the same page and so the
reader (still reading that page) would need to do mych more retry-
reading and so slow the whole stage down. Now with the aligned
batch sizes the reader doesn't contend with the writers in the
page cache.

Additionally the main store updating step have been split into steps
updating entities and properties separately, this to have the
entity updating able to go even faster.

The net result of this change as a whole should be that more often
the disk is the only main bottleneck. On test machines and development
laptops a 2x-3x performance improvement of the importer have
been observed.

tinwelint committed:
More aggressive assignment of processors earlier

which works better on machines with a lot of processors. Previously
the slow assignment would delay the full potential of the import since
all system resources wouldn't be fully utilized until a certain amount
of time into the import.

Also lowered the moving average size since now batch sizes often are
bigger (aligning with page size). This has the effect of snappier
and more changing average stats, but also a more responsive dynamic processor
assignments which means that a bottleneck step will be detected sooner
and assigned more processing power.

tinwelint committed:
Includes more information in overview print

of the import tool:
- number of processors in overview
- peak memory usage (from caching) during the import

tinwelint committed:
Configurable page size in importer

MishaDemianenko commented:
@tinwelint ping, broken atm

MishaDemianenko commented:
@tinwelint still broken on upgrade tests

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/NodeRelationshipCache.java:
```diff
@@ -410,7 +412,7 @@ public int getCount( long nodeId, Direction direction )
             this.base = base;
             assert chunkSize > 0;
-            this.array = arrayFactory.newDynamicByteArray( chunkSize,
+            this.array = arrayFactory.newDynamicByteArray( 1_000_000,
```
it looks like we actually should use chunkSize here and it looks like that now it can only be equal to CHUNK_SIZE which is 1_000_000 anyway and no changes is needed here

tinwelint replied:
oh right, I can revert that.

The thing about calculating number of dense nodes in advance is that it'd be a worst-case figure. We import relationships per type and so there will be a fraction of the entire relationship count for each type, right? The number of relationship groups needed in cache at most is the highest number of dense nodes within one type. We could calculate the biggest group, but that would be like saying that all nodes would be dense in that group.  Perhaps we could calculate the biggest number of dense nodes as well... hmm, lemme get back to you on that.

tinwelint replied:
Perhaps I'm mistaken, but I don't think we can easily calculate this and if we use the worst-case figure many imports will allocate lots of memory it doesn't need.

MishaDemianenko replied:
ok, lets keep if for now

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/NodeRelationshipCache.java:
```diff
@@ -410,7 +412,7 @@ public int getCount( long nodeId, Direction direction )
             this.base = base;
             assert chunkSize > 0;
-            this.array = arrayFactory.newDynamicByteArray( chunkSize,
+            this.array = arrayFactory.newDynamicByteArray( 1_000_000,
```
Should we also try to calculate number of relationship groups in advance to avoid dynamic array here as well?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/OffHeapNumberArray.java:
```diff
@@ -52,7 +52,11 @@ public void close()
         if ( !closed )
         {
-            UnsafeUtil.free( address );
+            if ( length > 0 )
```
lovely, lets have a test for this case? when we creating OffHeapNumberArray of 0 size and closing it without and we do not fail trying to free 0 address

tinwelint replied:
not sure we can test for it. Actually this may simply be an issue with the added tracking of allocation/access which can be enabled by a JVM flag, `CHECK_NATIVE_ACCESS`. Since UnsafeUtil can "allocate" 0 then deallocating 0 should be OK as well, sort of.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/RelationshipTypeCheckerStep.java:
```diff
@@ -42,13 +40,13 @@
+    private static final Comparator<Map.Entry<Object,MutableLong>> SORT_BY_ID_DESC =
             (e1,e2) -> Integer.compare( (Integer)e2.getKey(), (Integer)e1.getKey() );
-    private final ConcurrentMap<Object,AtomicLong> allTypes = new ConcurrentHashMap<>();
+    private final Map<Thread,Map<Object,MutableLong>> typeCheckers = new ConcurrentHashMap<>();
```
i think it can be useful to construct CHM with initial capacity, load factory and concurrencyLevel since we now how many workers we will have right?

tinwelint replied:
We don't know before-hand and it may change throughout the import. Items in that global map will change only when changing number of workers. I.e. this ain't an issue.

MishaDemianenko replied:
oki doki

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/RelationshipTypeCheckerStep.java:
```diff
@@ -116,14 +133,19 @@ protected void done()
         return result.toArray();
     }
+
+    private static class MutableLong
```
We already have couple of sworn brother classes like MutableBoolean, MutableInteger etc. defined in org.neo4j.kernel.impl.util i think we need to move new MutableLong there and ideally give move all of those in some nice mutable package

lutovich replied:
Can we use `org.apache.commons.lang3.mutable.*`? They should be in kernel module classpath already.

MishaDemianenko replied:
i think we have that discussion already couple of times before and decision was that we do not want dependency from kernel to commons-lang or try to minimise it as much as possible, that's how our own mutable classes were created

lutovich replied:
`neo4j-io` module brings `commons-lang3` transitively, so all those classes are available in the classpath

MishaDemianenko replied:
yes that's were _try to minimise_ part get into action :)

tinwelint replied:
Will the commons-lang class, it has methods like `add` and `increment`, sort of nice :)

MishaDemianenko replied:
so you change your mind on that?) that super cool then - please remove all other mutables as well :)

tinwelint replied:
that's better done as a separate PR to not pollute this change. We don't have to go all in on everything always

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/RelationshipTypeCheckerStep.java:
```diff
@@ -60,16 +58,21 @@ public RelationshipTypeCheckerStep( StageControl control, Configuration config,
     @Override
     protected void process( Batch<InputRelationship,RelationshipRecord> batch, BatchSender sender ) throws Throwable
     {
+        Map<Object,MutableLong> types = typeCheckers.get( Thread.currentThread() );
```
All those things can be shorted a bit to:

```
       Thread thread = Thread.currentThread();
        typeCheckers.computeIfAbsent( thread, ( t ) -> new HashMap<>() );

        Map<Object,MutableLong> typeMap = typeCheckers.get( thread );
        Stream.of( batch.input )
              .map( InputRelationship::typeAsObject )
              .forEach( type -> typeMap.computeIfAbsent( type, ( t ) -> new MutableLong() ).value++ );
        sender.send( batch );
```

tinwelint replied:
Right, cool. Even:

```
        Map<Object,MutableLong> typeMap = typeCheckers.computeIfAbsent( Thread.currentThread(), ( t ) -> new HashMap<>() );

        Stream.of( batch.input )
              .map( InputRelationship::typeAsObject )
              .forEach( type -> typeMap.computeIfAbsent( type, ( t ) -> new MutableLong() ).value++ );
        sender.send( batch );
```

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/RelationshipTypeCheckerStep.java:
```diff
@@ -60,16 +58,21 @@ public RelationshipTypeCheckerStep( StageControl control, Configuration config,
     @Override
     protected void process( Batch<InputRelationship,RelationshipRecord> batch, BatchSender sender ) throws Throwable
     {
+        Map<Object,MutableLong> types = typeCheckers.get( Thread.currentThread() );
```
functions can be of cause extracted as constants;
or at least we can try to use map.computeIfAbsent or map.putIfAbsent to remove those if checks

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/RelationshipTypeCheckerStep.java:
```diff
@@ -78,10 +81,24 @@ protected void process( Batch<InputRelationship,RelationshipRecord> batch, Batch
     protected void done()
     {
-        sortedTypes = allTypes.entrySet().toArray( new Map.Entry[allTypes.size()] );
+        Map<Object,MutableLong> mergedTypes = new HashMap<>();
```
the same comment as above except code example, of course :)

tinwelint replied:
yup, changed as well

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/staging/TicketedProcessing.java:
```diff
@@ -86,7 +86,7 @@ public boolean test( long ticket )
         {
             long queued = submittedTicket.get() - processedTicket.get();
-            return queued <= executor.processors( 0 );
+            return queued <= executor.processors( 0 ) | executor.isShutdown();
```
what was the race here? can we have some sort of test for that case?

tinwelint replied:
I forgot and trying to reproduce failed... all I know is that there was a 50/50 chance normal shutdown (or panic, dunno) hung the importer and that the added condition removed that issue.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/staging/DynamicProcessorAssigner.java:
```diff
@@ -81,7 +76,7 @@ public void check( StageExecution[] executions )
                 }
                 // Be a little more conservative removing processors from too fast steps
-                if ( removeProcessorFromPotentialIdleStep( execution ) )
+                if ( permits == 0 && removeProcessorFromPotentialIdleStep( execution ) )
```
can we please have a new test for new behaviour of permits evaluation

tinwelint replied:
I believe there were tests changed for this, tests that already set up such a scenario and which now behave differently.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/AssignRelationshipIdBatchStep.java:
```diff
@@ -0,0 +1,66 @@
+    protected void process( Batch<InputRelationship,RelationshipRecord> batch, BatchSender sender ) throws Throwable
+    {
+        // Assign first record id and send
+        batch.firstRecordId = nextId;
```
we do not perform validation of ids (for special IdGeneratorImpl.INTEGER_MINUS_ONE for a first batch), currently it's possible that first batch will contain reserved id.

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/CalculateDenseNodesStep.java:
```diff
@@ -19,43 +19,63 @@
+        if ( nodeId == -1 )
         {
-            for ( long id : ids )
+            if ( id == 0 )
```
lets add a constant for specific processor id and use it everywhere where we check for it

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/CalculateDenseNodesStep.java:
```diff
@@ -19,43 +19,63 @@
+            long endNodeId = batch.ids[idIndex++];
+            processNodeId( id, processors, startNodeId, relationship, relationship.startNode() );
+            if ( startNodeId != endNodeId ||            // avoid counting loops twice
+                 startNodeId == -1 || endNodeId == -1 ) // although always collect bad relationships
```
we do not need last or part 'endNodeId == -1'

MishaDemianenko replied:
maybe we can use null reference constant here for -1?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/CalculateRelationshipsStep.java:
```diff
@@ -26,6 +26,10 @@
 
+/**
+ * Keeps track of number of relationships to import, this to set highId in relationship store before import.
+ * This is because of the way double-unit records works, so the secondary units will end up beyond this limit.
```
`maxSpecific` field in this class can go....

tinwelint replied:
gone

MishaDemianenko commented on community/kernel/src/test/java/org/neo4j/unsafe/impl/batchimport/CapturingStep.java:
```diff
@@ -0,0 +1,51 @@
+        super( control, name, config, 1, additionalStatsProvider );
+    }
+
+    public Iterable<T> receivedBatches()
```
never used

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/staging/ForkedProcessorStep.java:
```diff
@@ -0,0 +1,215 @@
+{
+    // used by forked processors to count down when they're done, so that main processing thread
+    // knows when they're all done
+    private final AtomicInteger doneSignal = new AtomicInteger();
```
it looks like `CountDownLatch` should be used instead here

tinwelint replied:
No because we'd have to create a new such instance for every batch, which would be unnecessary imho.

MishaDemianenko replied:
CyclicBarrier then so you can reuse it :) ?

tinwelint committed:
Added more javadocs on import stages

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/staging/ForkedProcessorStep.java:
```diff
@@ -0,0 +1,215 @@
+    // it's not volatile, but piggy-backs on globalTicket for that
+    private T currentBatch;
+    // this ticket helps coordinating with the forked processors
+    private long globalTicket;
```
should `globalTicket` be volatile?

tinwelint replied:
There's a comment in `process` of that class which explains that it isn't needed

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/staging/ForkedProcessorStep.java:
```diff
@@ -0,0 +1,215 @@
+    // changes will only be applied between processing batches as to not interfere
+    private volatile int processorCount = 1;
+    // forked processors can communicate errors via this variable
+    private Throwable error;
```
also error and ticketThread should be volatile right?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/ParallelBatchImporter.java:
```diff
@@ -156,52 +160,40 @@ public void doImport( Input input ) throws IOException
+
+            // Stage 2 -- calculate dense node threshold
+            CalculateDenseNodesStage calculateDenseNodesStage = new CalculateDenseNodesStage(
+                    withBatchSize( config, config.batchSize()*10 ),
```
mmm...some magic multipliers :)

tinwelint replied:
mmh, well, we want the batches in this step to be a bit bigger since it favors the way some steps are parallelized (ForkedProcessorStep and any overhead it might add)

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/ParallelBatchImporter.java:
```diff
@@ -156,52 +160,40 @@ public void doImport( Input input ) throws IOException
                     calculateDenseNodesStage.getRelationshipTypes( Long.MAX_VALUE ),
-                    // Is batch size a good measure for considering a group of relationships a minority?
-                    calculateDenseNodesStage.getRelationshipTypes( config.batchSize() ) );
+                    calculateDenseNodesStage.getRelationshipTypes( 100 ) );
```
and some other bit of magic

tinwelint replied:
This whole "batch insert" thing for the last couple of is a bit odd, it was introduced because we had a test highlighting this (creates 500 relationships, each of its own type) which took a long time to run after the initial per-type. Although there have been optimizations to the stages since so that the overhead of running the relationship import loop for small groups of relationships isn't as expensive anymore. Hmm, perhaps we should simply remove this thing. Will measure.

tinwelint replied:
Could we perhaps take that as a separate change, the change here is made from empirical observations and is in fact better than the `config.batchSIze()`. We could revisit that stage of the import as a whole later.

MishaDemianenko replied:
ok lets take it separate

tinwelint committed:
Uses commons-lang MutableLong

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/RecordIdIterator.java:
```diff
@@ -0,0 +1,182 @@
+        return backwards( store.getNumberOfReservedLowIds(), store.getHighId(), config );
+    }
+
+    static class Forwards extends PrimitiveLongBaseIterator implements RecordIdIterator
```
this static is redundant

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/RecordIdIterator.java:
```diff
@@ -0,0 +1,182 @@
+        }
+    }
+
+    static class Backwards extends PrimitiveLongBaseIterator implements RecordIdIterator
```
also this static is redundant

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/RecordIdIterator.java:
```diff
@@ -0,0 +1,182 @@
+
+        public Forwards( long lowIncluded, long highExcluded, Configuration config )
+        {
+            this.lowIncluded = this.nextId = lowIncluded;
```
can we please split those assignments

tinwelint replied:
that's a matter of taste, isn't it? But sure

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/RecordIdIterator.java:
```diff
@@ -0,0 +1,182 @@
+        {
+            this.lowIncluded = lowIncluded;
+            this.batchSize = config.batchSize();
+            this.highExcluded = this.nextId = this.nextRoofId = this.floorId = highExcluded;
```
can we please split those assignments

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/RecordIdIterator.java:
```diff
@@ -0,0 +1,182 @@
+                else
+                {
+                    nextRoofId = floorId;
+                    nextId = floorId = floorId - min( size, batchSize );
```
can we please split those assignments

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/RelationshipEncoderStep.java:
```diff
@@ -39,78 +35,49 @@
+            RelationshipRecord relationship = batch.records[i];
+            long startNode = relationship.getFirstNode();
+            long endNode = relationship.getSecondNode();
+            if ( startNode == -1 || endNode == -1 )
```
can me maybe use empty reference constant or similar instead of -1?

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/RelationshipLinkbackStep.java:
```diff
@@ -0,0 +1,139 @@
+                {
+                    long prevRel = cache.getAndPutRelationship( record.getFirstNode(),
+                            Direction.BOTH, record.getId(), false );
+                    if ( prevRel == -1 )
```
magic -1's

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/RelationshipRecordPreparationStep.java:
```diff
@@ -0,0 +1,80 @@
+            InputRelationship batchRelationship = batch.input[i];
+            long startNodeId = batch.ids[idIndex++];
+            long endNodeId = batch.ids[idIndex++];
+            if ( startNodeId == -1 || endNodeId == -1 )
```
those guys again

MishaDemianenko commented on community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/staging/DynamicProcessorAssigner.java:
```diff
@@ -98,15 +93,15 @@ private int assignProcessorsToPotentialBottleNeck( StageExecution execution, int
              batchesPassedSinceLastChange( bottleNeckStep, doneBatches ) >= config.movingAverageSize() )
         {
+            // Assign 1/10th of the remaining permits. This will have processors being assigned more
+            // aggressively in the beginning of the run
```
that will also mean that in the end when will have less then 10 permits we will always use optimalProcessorIncrement (since it's hard to say what will be result of bottleNeck.other().floatValue() ) i would assume that `optimalProcessorIncrement` in those cases will be always 1, and that will cause a bit of throttling Is that correct?

tinwelint replied:
It's not about 10. Say we have 100 permits. The first time 10 will be assigned, the next assignment will be 9 since there are now only 90 left. Then 8 and so forth. `optimalProcessorIncrement` is the optimal amount of processors to assign given the difference between the bottleneck step and the next-slowest. Of course the next slowest may also be parallelizable and so they would previously add one or two every now and then, slowly getting faster and faster than the other, so to reach optimal setup took a while. This speeds up the process of getting to the optimal setup, with a slight loss of perfect precision, which wasn't really necessary anyway.

MishaDemianenko replied:
ok

MishaDemianenko commented:
@tinwelint a lot of places in batch importer use directly java System class to get current milliseconds or nanos, maybe we can switch it to use Clocks instead?
But this is absolutely independent from this pr, just one more thing that i notice.

MishaDemianenko commented:
@tinwelint first review completed, couple of comments most of them are minors

tinwelint committed:
Checks reserved ids for first relationship batch

tinwelint committed:
Fixes an issue in ForkedProcessorStep

with spurious wake-ups potentially hanging the processing, also tests for it.
Cleaned up volatile variables and clarifies where correct memory barriers are.

tinwelint committed:
Code cleanup, like using constant for -1

tinwelint commented:
@MishaDemianenko most comments fixed (resulting in some new commits)

alexaverbuch commented:
Import ran and was faster.
As we don't track import times I don't have exact numbers, but it took 6hours and that is up to 2x faster than I remember it taking in the past.

```
IMPORT DONE in 5h 59m 33s 963ms.

Imported:
  287854005 nodes
  2095300858 relationships
  2295312288 properties
Peak memory usage: 15.55 GB
2016-08-30 17:36:21.676+0000 INFO  [o.n.u.i.b.ParallelBatchImporter] Import completed, took 5h 59m 33s 963ms. Imported:
  287854005 nodes
  2095300858 relationships
  2295312288 properties
2016-08-30 17:36:21.923+0000 INFO  [o.n.k.i.s.c.CountsTracker] About to rotate counts store at transaction 1 to [/mnt/pcie-ssd/ldbc/db_sf100_p064_dense1_utc_day_30ce_v3/neostore.counts.db.b], from [/mnt/pcie-ssd/ldbc/db_sf100_p064_dense1_utc_day_30ce_v3/neostore.counts.db.a].
2016-08-30 17:36:22.040+0000 INFO  [o.n.k.i.s.c.CountsTracker] Successfully rotated counts store at transaction 1 to [/mnt/pcie-ssd/ldbc/db_sf100_p064_dense1_utc_day_30ce_v3/neostore.counts.db.b], from [/mnt/pcie-ssd/ldbc/db_sf100_p064_dense1_utc_day_30ce_v3/neostore.counts.db.a].
Data imported in: 359 min, 41 sec
LdbcSnbImporterParallelDense1  Creating Indexes & Constraints
LdbcIndexer  No Node Mandatory Constraints To Create
LdbcIndexer  No Relationship Mandatory Constraints To Create
LdbcIndexer  No Unique Constraints To Create
LdbcIndexer  Creating Schema Indexes For
LdbcIndexer     (Forum , id)
LdbcIndexer     (Country , name)
LdbcIndexer     (Person , firstName)
LdbcIndexer     (TagClass , name)
LdbcIndexer     (Person , id)
LdbcIndexer     (City , name)
LdbcIndexer     (Tag , id)
LdbcIndexer     (Message , id)
LdbcIndexer     (GraphMetaData , id)
LdbcIndexer     (Country , id)
LdbcIndexer     (Tag , name)
LdbcIndexer  Waiting For Indexes To Build
LdbcIndexer     (Forum , [id]) - POPULATING
LdbcIndexer     (Forum , [id]) - ONLINE
LdbcIndexer     (TagClass , [name]) - ONLINE
LdbcIndexer     (Tag , [id]) - ONLINE
LdbcIndexer     (Tag , [name]) - ONLINE
LdbcIndexer     (City , [name]) - ONLINE
LdbcIndexer     (GraphMetaData , [id]) - ONLINE
LdbcIndexer     (Country , [name]) - ONLINE
LdbcIndexer     (Country , [id]) - ONLINE
LdbcIndexer     (Person , [firstName]) - ONLINE
LdbcIndexer     (Person , [id]) - ONLINE
LdbcIndexer     (Message , [id]) - POPULATING
LdbcIndexer     (Message , [id]) - ONLINE
Indexes built in: 27 min, 7 sec
Shutting down...Done
```

MishaDemianenko commented:
@tinwelint looks good will merge on green

tinwelint commented:
@alexaverbuch that's great, it's a shame thought that the import details is missing from your pasted output :)
